### DISCUSSION TITLE
[Fast Data] Loop consuming messages

### DISCUSSION THREAD

**Discussion by fabio-percivaldi** [mia-platform/community#195] [Q&A]
Hi Everyone, 

we noticed a strange behaviour in the Fast Data system, without releasing any changes on the cluster, from the 3rd of February the Realtime Updater started to consume the same messages creating a loop.

We are using Fast Data to consume csv files once per day and after the 3rd of February the RTU is accumulating messages every day, every you can see this behaviour in the dashboard
<img width="581" alt="image" src="https://user-images.githubusercontent.com/30175434/217546559-3df34200-bade-4586-858e-c94f9b4d60ba.png">

From the logs is possible to see some problems related to the upsert function, in particular we are hitting a "duplicated key error", i will try to summarize the behaviour:

1. the RTU consumes some messages from Kafka, the offset (__internal__counter) of the message is 1675386135304 (that is Friday, February 3, 2023 2:02:15.304 AM)
2. it searches for some projections that has no `__internal__counter` or that has an `__internal__counter` lower then or equal to 1675386135304
3. it does not find a document, because for some reason the document with the given id is updated with a `__internal__counter` that is **after** the `February 3, 2023 2:02:15.304 AM` 
4. it tries to insert the document inside the database
5. it gets a `E11000 duplicate key error collection` since the document already exists but with a more recent `__internal__counter`
6. the batch is totally discarded
<img width="2545" alt="Screenshot 2023-02-08 alle 11 58 07" src="https://user-images.githubusercontent.com/30175434/217546382-44239fe6-6f35-4f27-b7c4-20874b617e9c.png">

This seems to be happening only for 2 of the 3 partitions of the topic:
<img width="1119" alt="Screenshot 2023-02-08 alle 14 24 10" src="https://user-images.githubusercontent.com/30175434/217542575-e96756ec-bb68-4d4a-b499-02457270f2df.png">

**Comment by ianFar96** [comment]
Hi fabbio,

This mechanism is made on purpose so we don't update a record with a more recent update.

The problem here is the batch gets discarded so the RTU keeps iterating over the same batch. In other words, it does not skip the old message, right?

**Reply by fabio-percivaldi** [reply]
yes exactly, i agree that the batch should be discarded, but we should then seek to a most recent offset and actually discard the old messages (that i know it is the actual behaviour but for some reason it's not happening here)

**Reply by ianFar96** [reply]
Did you update the RTU the 3rd of february? what version were you using and what version you use now?

**Reply by fabio-percivaldi** [reply]
no we didn't update the RTU neither the projections or single view. We didn't release anything on the cluster at all. Anyway now we are using the version `7.1.1`

**Reply by ianFar96** [reply]
I see, I just noticed that it tries to insert a record with all the primary keys to null.
Fortunately I have good news, this is a bug we solved with the version `7.2.0` which is the latest.

**Reply by fabio-percivaldi** [reply]
> I see, I just noticed that it tries to insert a record with all the primary keys to null. Fortunately I have good news, this is a bug we solved with the version `7.2.0` which is the latest.

that indeed is a great news, i will try to update to the version `7.2.0`

**Reply by fabio-percivaldi** [reply]
@ianFar96 is it enough changing the version of the RTU or we should also do a offset reset or delete old messages or something else?

**Reply by ianFar96** [reply]
Should be enough just changing the version, it should skip the messages by itself.

**Reply by fabio-percivaldi** [reply]
Updating the version of the Realtime Updater fixed the issue even though we had some problem:
The service was trying to generate projections changes even if the messages he was consuming were old, in that case it's not necessary to generate projection changes in my opinion, we resolved the problem disabling projection changes generation and this action speed up the offset updates (otherwise the service would still be stuck)
